ccl/changefeedccl/avro: TestAvroSchema failed

ccl/changefeedccl/avro.TestAvroSchema [failed](https://mesolite.cluster.engflow.com/invocations/default/92cec908-2e8a-4699-b9a6-eeb05b8ef018?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2wvYXZybzphdnJvX3Rlc3Q=) on master @ [c3a8ba737b223d7e96a50b4cfeefbe743528374f](https://github.com/cockroachdb/cockroach/commits/c3a8ba737b223d7e96a50b4cfeefbe743528374f):

Fatal error:

```
panic: failed to parse locale "C": language: tag is not well-formed [recovered]
	panic: failed to parse locale "C": language: tag is not well-formed [recovered]
	panic: failed to parse locale "C": language: tag is not well-formed

```
Stack: 

```
goroutine 14 [running]:
testing.tRunner.func1.2({0x452a2c0, 0xc0010f1370})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x452a2c0?, 0xc0010f1370?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:132 +0x3aa
panic({0x452a2c0?, 0xc0010f1370?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatumWithNullChance(0xc001cad200, 0xc0012ca9c0, 0x0, 0x0, 0x0)
	pkg/sql/randgen/datum.go:296 +0x1d11
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatum(0xc0012ca9c0?, 0x7?, 0x50?)
	pkg/sql/randgen/datum.go:47 +0x2b
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.TestAvroSchema(0xc0013ec680)
	pkg/ccl/changefeedccl/avro/avro_test.go:363 +0x78c
testing.tRunner(0xc0013ec680, 0x4c2eb90)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAvroSchema
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAvroSchema504321186
    test_log_scope.go:76: use -show-logs to present logs inline
    datum.go:296: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestAvroSchema504321186
    datum.go:296: panic: failed to parse locale "C": language: tag is not well-formed
--- FAIL: TestAvroSchema (0.01s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAvroSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47184

ccl/changefeedccl/avro.TestAvroSchema [failed](https://tanzanite.cluster.engflow.com/invocations/default/8a7647cd-8d67-4f6b-828f-4755c65f4db4?testReportRun=7&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2wvYXZybzphdnJvX3Rlc3Q=) on master @ [d276833ff08ef11d7894b7b0124f1883d90a3377](https://github.com/cockroachdb/cockroach/commits/d276833ff08ef11d7894b7b0124f1883d90a3377):

Fatal error:

```
panic: failed to parse locale "C": language: tag is not well-formed [recovered]
	panic: failed to parse locale "C": language: tag is not well-formed [recovered]
	panic: failed to parse locale "C": language: tag is not well-formed

```
Stack: 

```
goroutine 14 [running]:
testing.tRunner.func1.2({0x452ad60, 0xc0010c38b0})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x452ad60?, 0xc0010c38b0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:132 +0x3aa
panic({0x452ad60?, 0xc0010c38b0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatumWithNullChance(0xc000adb4d0, 0xc001eb6780, 0x0, 0x0, 0x0)
	pkg/sql/randgen/datum.go:296 +0x1d11
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatum(0xc001eb6780?, 0x7?, 0x50?)
	pkg/sql/randgen/datum.go:47 +0x2b
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.TestAvroSchema(0xc001f84680)
	pkg/ccl/changefeedccl/avro/avro_test.go:363 +0x78c
testing.tRunner(0xc001f84680, 0x4c2f6d0)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAvroSchema
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAvroSchema4092901559
    test_log_scope.go:76: use -show-logs to present logs inline
    datum.go:296: -- test log scope end --
    datum.go:296: panic: failed to parse locale "C": language: tag is not well-formed
--- FAIL: TestAvroSchema (0.00s)

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestAvroSchema4092901559
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAvroSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl/avro.TestAvroSchema [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18744480?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18744480?buildTab=artifacts#/) on master @ [bce8eb6edbbe879d376ab0c99e9d1129e5afef96](https://github.com/cockroachdb/cockroach/commits/bce8eb6edbbe879d376ab0c99e9d1129e5afef96):


```
=== RUN   TestAvroSchema
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/7d61bee4fd5b3770bd1965d4940b9e8f/logTestAvroSchema435208134
    test_log_scope.go:76: use -show-logs to present logs inline
    datum.go:296: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /artifacts/tmp/_tmp/7d61bee4fd5b3770bd1965d4940b9e8f/logTestAvroSchema435208134
    datum.go:296: panic: failed to parse locale "POSIX": language: tag is not well-formed
--- FAIL: TestAvroSchema (0.01s)
panic: failed to parse locale "POSIX": language: tag is not well-formed [recovered]
	panic: failed to parse locale "POSIX": language: tag is not well-formed [recovered]
	panic: failed to parse locale "POSIX": language: tag is not well-formed

goroutine 36 [running]:
testing.tRunner.func1.2({0x3fbf9c0, 0x400241d0e0})
	GOROOT/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x33c
panic({0x3fbf9c0?, 0x400241d0e0?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:132 +0x384
panic({0x3fbf9c0?, 0x400241d0e0?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatumWithNullChance(0x4000035650, 0x4001d8cc00, 0x0, 0x0, 0x0)
	pkg/sql/randgen/datum.go:296 +0x1860
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatum(0x4001d8cc00?, 0x7?, 0x48?)
	pkg/sql/randgen/datum.go:47 +0x30
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.TestAvroSchema(0x400025d380)
	pkg/ccl/changefeedccl/avro/avro_test.go:363 +0x5e8
testing.tRunner(0x400025d380, 0x46c4518)
	GOROOT/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x318
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAvroSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl/avro.TestAvroSchema [failed](https://tanzanite.cluster.engflow.com/invocations/default/296dc362-7481-4b36-8548-a4c9fef8278b?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2wvYXZybzphdnJvX3Rlc3Q=) on master @ [871fb9ff40888d77373b2da06c991e7f13d82789](https://github.com/cockroachdb/cockroach/commits/871fb9ff40888d77373b2da06c991e7f13d82789):

Fatal error:

```
panic: failed to parse locale "default": language: tag is not well-formed [recovered]
	panic: failed to parse locale "default": language: tag is not well-formed [recovered]
	panic: failed to parse locale "default": language: tag is not well-formed

```
Stack: 

```
goroutine 14856 [running]:
testing.tRunner.func1.2({0x45339e0, 0xc00121d510})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x45339e0?, 0xc00121d510?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:132 +0x3aa
panic({0x45339e0?, 0xc00121d510?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatumWithNullChance(0xc001172f00, 0xc0024e4780, 0x0, 0x0, 0x0)
	pkg/sql/randgen/datum.go:296 +0x1d11
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatum(0xc0024e4780?, 0x7?, 0x50?)
	pkg/sql/randgen/datum.go:47 +0x2b
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.TestAvroSchema(0xc0018ec9c0)
	pkg/ccl/changefeedccl/avro/avro_test.go:363 +0x78c
testing.tRunner(0xc0018ec9c0, 0x4c37870)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAvroSchema
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAvroSchema2856634695
    test_log_scope.go:76: use -show-logs to present logs inline
    datum.go:296: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestAvroSchema2856634695
    datum.go:296: panic: failed to parse locale "default": language: tag is not well-formed
--- FAIL: TestAvroSchema (0.30s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAvroSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl/avro.TestAvroSchema [failed](https://tanzanite.cluster.engflow.com/invocations/default/aae51751-59c0-41e3-ba8f-ec7eabdcf020?testReportRun=4&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2wvYXZybzphdnJvX3Rlc3Q=) on master @ [bce8eb6edbbe879d376ab0c99e9d1129e5afef96](https://github.com/cockroachdb/cockroach/commits/bce8eb6edbbe879d376ab0c99e9d1129e5afef96):

Fatal error:

```
panic: failed to parse locale "C": language: tag is not well-formed [recovered]
	panic: failed to parse locale "C": language: tag is not well-formed [recovered]
	panic: failed to parse locale "C": language: tag is not well-formed

```
Stack: 

```
goroutine 14 [running]:
testing.tRunner.func1.2({0x452c860, 0xc00099c8b0})
	GOROOT/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x377
panic({0x452c860?, 0xc00099c8b0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:132 +0x3aa
panic({0x452c860?, 0xc00099c8b0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatumWithNullChance(0xc000f77e00, 0xc001e7e9c0, 0x0, 0x0, 0x0)
	pkg/sql/randgen/datum.go:296 +0x1d11
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatum(0xc001e7e9c0?, 0x7?, 0x50?)
	pkg/sql/randgen/datum.go:47 +0x2b
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.TestAvroSchema(0xc001072680)
	pkg/ccl/changefeedccl/avro/avro_test.go:363 +0x78c
testing.tRunner(0xc001072680, 0x4c31330)
	GOROOT/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAvroSchema
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAvroSchema420067596
    test_log_scope.go:76: use -show-logs to present logs inline
    datum.go:296: -- test log scope end --
    datum.go:296: panic: failed to parse locale "C": language: tag is not well-formed
--- FAIL: TestAvroSchema (0.01s)

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestAvroSchema420067596
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAvroSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl/avro.TestAvroSchema [failed](https://tanzanite.cluster.engflow.com/invocations/default/2787df0b-2c01-4364-b2e5-745baa9b5e58?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2wvYXZybzphdnJvX3Rlc3Q=) on master @ [871fb9ff40888d77373b2da06c991e7f13d82789](https://github.com/cockroachdb/cockroach/commits/871fb9ff40888d77373b2da06c991e7f13d82789):

Fatal error:

```
panic: failed to parse locale "POSIX": language: tag is not well-formed [recovered]
	panic: failed to parse locale "POSIX": language: tag is not well-formed [recovered]
	panic: failed to parse locale "POSIX": language: tag is not well-formed

```
Stack: 

```
goroutine 14 [running]:
testing.tRunner.func1.2({0x6f4f820, 0xc000fc7240})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0x6f4f820?, 0xc000fc7240?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:132 +0x51a
panic({0x6f4f820?, 0xc000fc7240?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatumWithNullChance(0xc000d71d70, 0xc0017a8780, 0x0, 0x0, 0x0)
	pkg/sql/randgen/datum.go:296 +0x2bc5
github.com/cockroachdb/cockroach/pkg/sql/randgen.RandDatum(0xc000d71d70, 0xc0017a8780, 0x0)
	pkg/sql/randgen/datum.go:47 +0x52
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/avro.TestAvroSchema(0xc001c82820)
	pkg/ccl/changefeedccl/avro/avro_test.go:363 +0xaae
testing.tRunner(0xc001c82820, 0x765a0f0)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestAvroSchema
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestAvroSchema3003204320
    test_log_scope.go:76: use -show-logs to present logs inline
    datum.go:296: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestAvroSchema3003204320
    datum.go:296: panic: failed to parse locale "POSIX": language: tag is not well-formed
--- FAIL: TestAvroSchema (0.10s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAvroSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/avro/avro_test.go
+++ b/pkg/ccl/changefeedccl/avro/avro_test.go
@@ -342,7 +342,14 @@ func TestAvroSchema(t *testing.T) {
 		switch typ.Family() {
 		case types.StringFamily:
 			collationTags := collatedstring.Supported()
+			// "C" and "POSIX" locales are not allowed for collated string
+			// columns in CRDB (see collatedstring logic tests),
+			// so we don't expect these types to be emitted by changefeeds.
 			randCollationTag := collationTags[rand.Intn(len(collationTags))]
+			for randCollationTag == collatedstring.CCollationTag ||
+				randCollationTag == collatedstring.PosixCollationTag {
+				randCollationTag = collationTags[rand.Intn(len(collationTags))]
+			}
 			collatedType := types.MakeCollatedString(typ, randCollationTag)
 			typesToTest = append(typesToTest, collatedType)
 		}

